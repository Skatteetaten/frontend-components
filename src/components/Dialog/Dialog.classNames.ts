import { getTheme } from '@uifabric/styling';
import { mergeStyleSets } from '@uifabric/merge-styles';
import { FontSizes, FontWeights } from '../utils/fonts';
import { PaletteProps } from '..';

const dekor = require('./assets/footerDekor.svg');
const logo = require('./assets/ske-logo.svg');
// @ts-ignore TODO
function setMinMaxWidth(props) {
  return {
    minWidth: props.minWidth,
    maxWidth: props.maxWidth
  };
}
// @ts-ignore TODO
function getMainBackgroundStyle(props) {
  const palette = getTheme().palette as PaletteProps;

  if (props.layoutStyle === 'important') {
    return {
      background: `url(${dekor})`,
      backgroundSize: 'calc(100% + 1px) 20px',
      backgroundRepeat: 'no-repeat',
      backgroundPosition: 'bottom left',
      paddingBottom: 30,
      backgroundColor: palette.white,
      border: 'none !important'
    };
  } else {
    return {};
  }
}
// @ts-ignore TODO
function getHeaderBackgroundStyle(props) {
  if (props.layoutStyle === 'important') {
    return {
      backgroundImage: `url(${logo})`,
      backgroundSize: '40px 100%',
      backgroundRepeat: 'no-repeat',
      backgroundPosition: 'top -30px left 20px',
      paddingTop: 70
    };
  } else {
    return {};
  }
}
// @ts-ignore TODO
export var getClassNames = function getClassNames(props) {
  const palette = getTheme().palette as PaletteProps;

  return mergeStyleSets({
    main: {
      displayName: `SkeDialog`,
      position: 'absolute',
      selectors: {
        '& .ms-Dialog-main': {
          selectors: {
            '@media (min-width: 480px)': {
              ...setMinMaxWidth(props),
              borderColor: palette.skeColor.burgundy,
              borderStyle: 'solid',
              borderWidth: '4px'
            }
          },
          ...getMainBackgroundStyle(props)
        },
        '& .ms-Dialog-header': {
          ...getHeaderBackgroundStyle(props)
        },
        '& .ms-Button.ms-Dialog-button--close i': {
          fontWeight: FontWeights.bold,
          fontSize: FontSizes.icon,
          opacity: 1
        },
        '& .ms-Button.ms-Dialog-button--close:hover i': {
          fontWeight: FontWeights.bold,
          fontSize: FontSizes.icon,
          opacity: 0.7,
          transition: 'opacity 300ms ease-out'
        },
        '& .ms-Dialog-title': {
          fontSize: FontSizes.xLarge,
          fontWeight: FontWeights.semibold,
          padding: props.layoutStyle === 'airy' ? '128px 64px 0 64px' : 20
        },
        '& .ms-Dialog-subText': {
          fontSize: FontSizes.medium
        },
        '& .ms-Overlay': {
          backgroundColor: 'rgba(255, 255, 255, 0.8)'
        },
        '& .ms-Dialog-inner': {
          padding:
            props.layoutStyle === 'airy' ? '0 64px 64px ' : '0 20px 20px 20px'
        }
      }
    },
    helpButton: {
      displayName: 'SkeHelpButton',
      zIndex: 100,
      position: 'absolute',
      bottom: '-60px',
      left: '-15px'
    }
  });
};
