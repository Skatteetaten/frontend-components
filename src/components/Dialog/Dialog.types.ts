import { IDialogProps } from '@fluentui/react';
import { Language } from '../utils/Language';

export interface DialogProps extends IDialogProps {
  /** doNotLayer prop for den help-Callout */
  doNotLayer?: boolean;
  isModeless?: boolean;
  /** Language selection for what the screen reader reads out. Default is Norwegian Bokm책l */
  language?: Language;
  /** Om dialog skal ha mer padding for et luftigere uttrykk */
  layoutStyle?: 'normal' | 'airy' | 'important';
  /** Om det er s책 mye innhold at det g책r over flere "sider" (fikser scroll inni dialog p책 ipad) */
  tabletContentOverflows?: boolean;
  /** Ventevarsel */
  waitingWarning?: boolean;
  /** Ventevarsel knapp tekst */
  waitingWarningBtnText?: string;
}
export type DialogState = {
  isCalloutVisible: boolean;
};
