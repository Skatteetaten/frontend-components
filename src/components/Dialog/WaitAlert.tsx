import * as React from 'react';
import waitAlertSvg from './assets/wait_alert_illustration.svg';
import { t } from '../utils';
import { Button } from '../Button';

const WaitAlert: React.FC<{
  onDismiss?: (ev?: React.MouseEvent<HTMLButtonElement>) => any;
  waitAlertBtnText?: string;
}> = (props) => {
  return (
    <div className={'wait-alert-content'}>
      <img
        className={'wait-alert-svg'}
        src={waitAlertSvg}
        alt={t('dialog.waitAlert.alt')}
      />
      {props.children ? (
        props.children
      ) : (
        <>
          <div
            role="heading"
            id={'waitAlertHeading'}
            className={'wait-alert-heading'}
            aria-level={1}
          >
            {t('dialog.waitAlert.title')}
          </div>
          <p id={'waitAlertText'}>{t('dialog.waitAlert.paragraph')}</p>
        </>
      )}
      <Button
        buttonStyle="primary"
        className={'wait-alert-btn'}
        onClick={() => {
          props.onDismiss && props.onDismiss();
        }}
      >
        {props.waitAlertBtnText
          ? props.waitAlertBtnText
          : t('dialog.waitAlert.button_text')}
      </Button>
    </div>
  );
};

export default WaitAlert;
