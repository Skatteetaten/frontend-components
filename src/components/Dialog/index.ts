export { Dialog as default } from './Dialog';
export * from './Dialog';
