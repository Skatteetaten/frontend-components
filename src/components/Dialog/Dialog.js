import React from 'react';
import PropTypes from 'prop-types';
import { getClassNames } from './Dialog.classNames';
import Callout from '../Callout/Callout';
import classnames from 'classnames';

import {
  DialogType,
  DialogFooter,
  Dialog as FabricDialog
} from 'office-ui-fabric-react/lib-commonjs/Dialog';

import IconButton from '../IconButton/IconButton';

/**
 * @visibleName Dialog (Dialogboks)
 */
export default class Dialog extends React.PureComponent {
  static Footer = DialogFooter;
  static Type = DialogType;

  static propTypes = {
    /** Om dialogen er skjult eller ikke */
    hidden: PropTypes.bool,
    /** Callback til n√•r dialogen lukkes */
    onDismiss: PropTypes.func,
    /** Tittel som skal vises i dialogen */
    title: PropTypes.string,
    /** Undertekst som skal vises i dialogen */
    subText: PropTypes.string,
    /** Hvorvidt om dialogen skal blokkere for annen brukerinput */
    isBlocking: PropTypes.bool,
    /** Type dialog; normal eller stor header  */
    type: PropTypes.oneOf([Dialog.Type.normal, Dialog.Type.largeHeader]),
    /** Om dialog skal ha mer padding for et luftigere uttrykk */
    layoutStyle: PropTypes.oneOf(['normal', 'airy', 'important']),
    /** Hjelpetekst som skal vises i Callout */
    helpText: PropTypes.string,
    /** Default max-width for the dialog box */
    dialogMinWidth: PropTypes.string,
    /** Default min-width for the dialog box */
    dialogMaxMinWidth: PropTypes.string,
    className: PropTypes.string
  };
  // TODO: Use Fabric props: dialogDefaultMinWidth / dialogDefaultMaxWidth
  // when bug issue is resolved (https://github.com/OfficeDev/office-ui-fabric-react/issues/5175)

  static defaultProps = {
    type: Dialog.Type.normal,
    layoutStyle: 'normal'
  };

  constructor(props) {
    super(props);
    this.state = {
      isCalloutVisible: false
    };
    this._onClick = this._onClick.bind(this);
    this._onDismiss = this._onDismiss.bind(this);
  }

  render() {
    const {
      children,
      hidden,
      onDismiss,
      title,
      subText,
      isBlocking,
      type,
      helpText,
      dialogMinWidth,
      dialogMaxWidth,
      className,
      layoutStyle,
      ...props
    } = this.props;
    const styles = getClassNames(this.props);
    let { isCalloutVisible } = this.state;

    return (
      <div>
        <FabricDialog
          {...props}
          role="dialog"
          aria-label={title}
          hidden={hidden}
          onDismiss={onDismiss}
          minWidth={dialogMinWidth}
          maxWidth={dialogMaxWidth}
          dialogContentProps={{
            type: type,
            title,
            subText,
            closeButtonAriaLabel: 'Lukk'
          }}
          modalProps={{
            isBlocking,
            className: classnames(styles.main, className)
          }}
        >
          {helpText && (
            <span
              className={styles.helpButton}
              ref={menuButton => (this._iconButtonElement = menuButton)}
            >
              <IconButton
                buttonType={'large'}
                icon={'HelpOutline'}
                iconProps={{ iconName: 'Calendar' }}
                onClick={this._onClick}
                title="Hjelp"
                aria-label="Hjelp"
              />
            </span>
          )}
          {isCalloutVisible && (
            <Callout
              directionalHint={Callout.POS_TOP_LEFT}
              color={Callout.HELP}
              ariaLabel={'Hjelpetekst'}
              target={this._iconButtonElement}
              onClose={this._onDismiss}
            >
              <p>{helpText}</p>
            </Callout>
          )}
          {children}
        </FabricDialog>
      </div>
    );
  }

  _onClick() {
    this.setState({
      isCalloutVisible: !this.state.isCalloutVisible
    });
  }

  _onDismiss() {
    this.setState({
      isCalloutVisible: false
    });
  }
}
