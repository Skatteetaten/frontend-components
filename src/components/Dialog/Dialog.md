** Dialog brukes til å vise innhold midt på skjermen og tiltrekke seg brukerens oppmerksomhet. **

```js
import Dialog from '@skatteetaten/frontend-components/Dialog';
import Button from '@skatteetaten/frontend-components/Button';
import ActionButton from '@skatteetaten/frontend-components/ActionButton';

const initialState = { hideDialog: true };

function closeDialog() {
  setState({ hideDialog: true });
}

<div>
  <Button
    type="secondary"
    aria-haspopup="true"
    onClick={() => setState({ hideDialog: false })}
  >
    Vis standard dialog
  </Button>
  <Dialog
    id="hipp"
    hidden={state.hideDialog}
    type={Dialog.Type.normal}
    onDismiss={closeDialog}
    title="Kansellere arbeidsoppgaven?"
    dialogMinWidth="400px"
    dialogMaxWidth="600px"
  >
    <p>Er du sikker på at du vil kansellere arbeidsoppgaven?</p>

    <p>Handlingen kan ikke reverseres</p>
    <Dialog.Footer>
      <ActionButton onClick={closeDialog}>Avbryt</ActionButton>
      <Button onClick={closeDialog} hoved>
        Kanseller
      </Button>l
    </Dialog.Footer>
  </Dialog>
</div>;
```

```js
import Dialog from '@skatteetaten/frontend-components/Dialog';
import Button from '@skatteetaten/frontend-components/Button';
import NavigationTile from '@skatteetaten/frontend-components/NavigationTile';

const initialState = { hideDialog: true };

function closeDialog() {
  setState({ hideDialog: true });
}

const content1 = [
  {
    to: '#',
    title: 'Næringsrapport skatt AS',
    icon: 'ArrowForward'
  },
  {
    to: '#',
    title: 'Næringsrapport skatt ENK',
    icon: 'ArrowForward'
  }
];

<div>
  <Button
    type="secondary"
    aria-haspopup="true"
    onClick={() => setState({ hideDialog: false })}
  >
    Vis luftig dialog
  </Button>
  <Dialog
    hidden={state.hideDialog}
    type={Dialog.Type.normal}
    onDismiss={closeDialog}
    title="Velg den inngangen som passer for deg"
    dialogMinWidth="500px"
    dialogMaxWidth="600px"
    layoutStyle={'airy'}
  >
    <p>
      Løsningen for aksjeselskap (AS) åpner i februar mens den blir tilgjengelig
      for enkeltpersonforetak (ENK) i april
    </p>
    <br />
    <NavigationTile
      naviStyle="left"
      naviIcon="left"
      naviTitle="left"
      naviDesc="left"
      contents={content1}
      renderContent={(to, content1) => <a href={'#'}>{content1}</a>}
    />
  </Dialog>
</div>;
```

```js
import Dialog from '@skatteetaten/frontend-components/Dialog';
import Button from '@skatteetaten/frontend-components/Button';
import ActionButton from '@skatteetaten/frontend-components/ActionButton';

const initialState = { hideDialog: true };

function closeDialog() {
  setState({ hideDialog: true });
}

<div>
  <Button
    type="secondary"
    aria-haspopup="true"
    onClick={() => setState({ hideDialog: false })}
  >
    Vis viktig dialog
  </Button>
  <Dialog
    id="hipp"
    hidden={state.hideDialog}
    type={Dialog.Type.normal}
    layoutStyle={'important'}
    onDismiss={closeDialog}
    title="Viktig melding!"
    dialogMinWidth="400px"
    dialogMaxWidth="600px"
  >
    <p>Løsning er ikke kommet i drift ennå eller tatt ned for vedlikehold</p>
    <Dialog.Footer>
      <ActionButton onClick={closeDialog}>
        Les mer på skatteetaten.no
      </ActionButton>
    </Dialog.Footer>
  </Dialog>
</div>;
```

```js noeditor
import Accordion from '@skatteetaten/frontend-components/Accordion';
import AccordionItem from '@skatteetaten/frontend-components/Accordion/AccordionItem';
<Accordion>
  <AccordionItem
    toggleContent
    isOpen
    toggleButtonText={'Bruk'}
    stepId={'step-1-1'}
  >
    <p>
      En dialogboks (modal) brukes gjerne til å gi brukeren et valg samtidig som
      konteksten beholdes.
    </p>
    <p>
      Hvis brukeren skal ta stilling til et valg i dialogen bør den settes opp
      slik at annen brukerinput blir blokkert.{' '}
    </p>
    <p>
      Dialogen finnes i tre varianter; «standard», «luftig» eller «viktig»:
      Luftig kan brukes hvor du ønsker å at dialogen skal tiltrekke seg ektra
      oppmerksomhet, mens viktig brukes dersom vi ønsker å formidle en viktig
      melding der Skatteetaten er avsender.
    </p>
    <p>Dialogen kan inneholde tekst, inputfelt, hjelpeikoner osv.</p>
    <p>
      Se{' '}
      <a href="https://www.skatteetaten.no/stilogtone/skrive/">
        Skatteetatens stil og tone
      </a>{' '}
      for hjelp til å skrive gode tekster.
    </p>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Universell utforming'}
    stepId={'step-1-2'}
  >
    <ul>
      <li>
        Bruk <b>aria-haspopup</b> for å indikere at en dialog åpnes ved klikk på
        en knapp.{' '}
      </li>
      <li>
        Sjekk at dialogen får fokus etter at den åpnes. Dette gjør det enkelere
        for en skjermleser å oppdage og lese opp innholdet.
      </li>
    </ul>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Tekniske detaljer'}
    stepId={'step-1-3'}
  >
    <p>
      <a
        href="https://developer.microsoft.com/en-us/fabric#/components/dialog#Implementation"
        target="_blank"
      >
        Se flere tilgjengelige props i Fabric dokumentasjonen
      </a>
    </p>
  </AccordionItem>
</Accordion>;
```
