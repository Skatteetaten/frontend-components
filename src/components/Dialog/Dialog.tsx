import classnames from 'classnames';
import * as React from 'react';
import Callout from '../Callout/Callout';
import { getClassNames } from './Dialog.classNames';
import {
  DialogType,
  DialogFooter,
  Dialog as FabricDialog,
  IDialogProps
} from 'office-ui-fabric-react/lib-commonjs/Dialog';
import IconButton from '../IconButton/IconButton';

interface DialogProps extends IDialogProps {
  /** Om dialog skal ha mer padding for et luftigere uttrykk */
  layoutStyle?: 'normal' | 'airy' | 'important';
  /** Hjelpetekst som skal vises i Callout */
  helpText?: string;
}
type DialogState = {
  isCalloutVisible: boolean;
};
/**
 * @visibleName Dialog (Dialogboks)
 */
export default class Dialog extends React.PureComponent<
  DialogProps,
  DialogState
> {
  static Footer = DialogFooter;
  static Type = DialogType;

  static defaultProps = {
    layoutStyle: 'normal',
    type: Dialog.Type.normal
  };
  // @ts-ignore TODO
  constructor(props) {
    super(props);
    this.state = {
      isCalloutVisible: false
    };
    this._onClick = this._onClick.bind(this);
    this._onDismiss = this._onDismiss.bind(this);
  }

  render() {
    const {
      children,
      title,
      subText,
      isBlocking,
      type,
      helpText,
      className,
      layoutStyle,
      ...props
    } = this.props;
    const styles = getClassNames(this.props);
    let { isCalloutVisible } = this.state;

    return (
      <div>
        {/*
        // @ts-ignore */}
        <FabricDialog
          {...props}
          role="dialog"
          dialogContentProps={{
            type: type,
            title,
            subText,
            closeButtonAriaLabel: 'Lukk'
          }}
          modalProps={{
            isBlocking,
            className: classnames(styles.main, className)
          }}
        >
          {helpText && (
            <span
              className={styles.helpButton}
              // @ts-ignore TODO
              ref={menuButton => (this._iconButtonElement = menuButton)}
            >
              <IconButton
                type={'large'}
                icon={'HelpOutline'}
                iconProps={{ iconName: 'Calendar' }}
                onClick={this._onClick}
                title="Hjelp"
                aria-label="Hjelp"
              />
            </span>
          )}
          {isCalloutVisible && (
            <Callout
              directionalHint={Callout.POS_TOP_LEFT}
              color={Callout.HELP}
              ariaLabel={'Hjelpetekst'}
              // @ts-ignore TODO
              target={this._iconButtonElement}
              onClose={this._onDismiss}
            >
              <p>{helpText}</p>
            </Callout>
          )}
          {children}
        </FabricDialog>
      </div>
    );
  }

  _onClick() {
    this.setState({
      isCalloutVisible: !this.state.isCalloutVisible
    });
  }

  _onDismiss() {
    this.setState({
      isCalloutVisible: false
    });
  }
}
