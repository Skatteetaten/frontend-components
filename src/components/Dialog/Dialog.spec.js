import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Dialog from './Dialog';

function oppsettShallow(props) {
  return shallow(<Dialog {...props} />);
}

function oppsettFullDOM(props) {
  return mount(<Dialog {...props} />);
}

describe('Dialog komponent', () => {
  it('matcher snapshot', () => {
    const wrapper = shallow(<Dialog />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('rendrer Dialog med riktige props', () => {
    const wrapper = oppsettFullDOM({
      layoutStyle: 'important',
      type: Dialog.Type.largeHeader,
      title: 'Tekst tittel',
      subText: 'Denne teksten vil vises i dialogen under tittelen',
      minWidth: '400px',
      maxWidth: '600px',
      hidden: false
    });

    const htmlElm = wrapper.find('StyledWithResponsiveMode').first();
    expect(wrapper.prop('layoutStyle')).toEqual('important');
    expect(wrapper.prop('type')).toEqual(1);
    expect(wrapper.find('.ms-Dialog-title').html()).toContain('Tekst tittel');
    expect(wrapper.find('.ms-Dialog-subText').html()).toContain(
      'Denne teksten vil vises i dialogen under tittelen'
    );

    expect(htmlElm.prop('minWidth')).toEqual('400px');
    expect(htmlElm.prop('maxWidth')).toEqual('600px');
  });
});
