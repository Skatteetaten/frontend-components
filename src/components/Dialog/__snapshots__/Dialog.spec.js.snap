// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog komponent matcher snapshot 1`] = `
<div>
  <StyledWithResponsiveMode
    dialogContentProps={
      Object {
        "closeButtonAriaLabel": "Lukk",
        "subText": undefined,
        "title": undefined,
        "type": 0,
      }
    }
    modalProps={
      Object {
        "className": "SkeDialog-40",
        "isBlocking": undefined,
      }
    }
    tabletContentOverflows={false}
  />
</div>
`;
